mesonbuild/compilers/c.py

```diff
@@ -93,7 +93,7 @@ def split_shlib_to_parts(self, fname):
     # The default behavior is this, override in MSVC
     def build_rpath_args(self, build_dir, from_dir, rpath_paths, build_rpath, install_rpath):
         if self.id == 'clang' and self.clang_type == compilers.CLANG_OSX:
-            return self.build_osx_rpath_args(build_dir, rpath_paths, build_rpath)
+            return [] # no rpath on macOS!
```

Comment:
Why are you doing this change here? It seems unrelated to this PR.

Reply:
No it is... the RPATH stuff is not necessary for macOS. If you want to link using relative paths, then you should be using `@executable_path` or `@loader_path` instead.

Reply:
This is not true, rpath does exist and is useful on macOS, it just works in a different way than it does on Linux.
> -rpath
> Add path to the runpath search path list for image being created.  At runtime, dyld uses the runpath when searching for dylibs whose load path begins with `@rpath/`.

And I still do not see how this is related to this PR.

Reply:
I didn't say it didn't exist on macOS 😉 

Basically what I am trying here is emulating libtool's behaviour, because I know libtool builds proper dylibs and proper GObject introspection bindings. Since libtool doesn't use rpath on macOS, neither am I. Also, if you are just doing a simple installation and will not move the installed files afterwards (which is true for the vast majority of cases, including MacPorts and Homebrew), then rpath is redundant anyway since the full path of the dylibs your executable or library depend on, will be hardcoded into the dylib header.

Reply:
> Since libtool doesn't use rpath on macOS, neither am I.

Yes thats true, still not really related to this PR ;)

> Also, if you are just doing a simple installation and will not move the installed files afterwards (which is true for the vast majority of cases, including MacPorts and Homebrew)

Yeah this might be true for some cases but not for all. For example for VLC I actually require the rpath stuff to be set properly, like it currently is done. I agree that this is debatable and probably there should be a way to configure this behavior as necessary and default to not have an rpath,  but thats exactly the reason why I think it should not just be changed here, ignoring the discussions about it in #3077 and #2121

Reply:
Note that I would be happy if a good solution for the rpath/install_name can be found, I just do not think it is in the scope of this PR and just removing rpath like you do here seems to break things, at least for my usecase. (Probably why the tests break too, no?)

Reply:
If you are going to be building a VLC app bundle (or any other app bundle or framework), then you will indeed need to modify the dylib headers accordingly using `install_name_tool`. But even then, the recommended way is to use `@executable_path` and `@loader_path` instead of `@rpath`.

Reply:
I would be happy to change the title of the PR to something like "Fix macOS dylibs and GObject introspection bindings" if you would like me to widen the scope 😄

Reply:
>  But even then, the recommended way is to use @executable_path and @loader_path instead of @rpath.

Why? Afaik from what I found, the recommended way on modern macOS versions is to go with the rpath approach and it works fine, better than what we had before and does not require us to use `install_name_tool` on all libraries, but only on the executable when bundling (to set the rpath correctly), making bundling much easier and less error-prone.

> I would be happy to change the title of the PR to something like "Fix macOS dylibs and GObject introspection bindings" if you would like me to widen the scope

It should not be done in this PR at all, I would argue. Just split up the commit and make a new PR with the rpath change? That way at least the library version stuff could be merged soon, hopefully? And we can continue the discussion about rpath on a dedicated PR about it and evaluate the solutions.

Reply:
> Why? Afaik from what I found, the recommended way on modern macOS versions is to go with the rpath approach and it works fine, better than what we had before and does not require us to use `install_name_tool` on all libraries, but only on the executable when bundling (to set the rpath correctly), making bundling much easier and less error-prone.

What you are saying is true for the binaries that are installed in the same meson build. That works well. But when the libraries are installed by a package manager (and we have no traces of meson files any longer), we end up with libraries ids using `@rpath`. When another package is built against that library, the binaries won't work because the library won't be found unless we hack the build system of each other build.

I believe that `@rpath` functionality should stay, it's useful, but there would have to be an alternative configuration option that would trigger `install_name_tool` to fix all paths at installation time.

Reply:
Yep, basically as I already said: In some cases (like when you have full control over both the binary and the libraries you want to link) rpath can be really useful, but in other cases (like the one you described) it is not, so we need to come up with a solution that suits these cases and probably a better default than what we have now. Which is why I am against having it in this PR, as I think there is no good reason to do this change together with adding library version information, as the two things are not related.

Reply:
Thanks a lot for the insight on the other case, in which a rpath approach would not work (or only with a lot of effort). I knew there were such cases but had no concrete idea how those look like.